[Impeller] Optionally support SamplerAddressMode::kDecal on the OpenGLES backend

ColdPaleLight opened the pull request:
fix https://github.com/flutter/flutter/issues/129358

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

jonahwilliams commented on impeller/renderer/backend/gles/sampler_gles.cc:
```diff
@@ -53,7 +53,8 @@ static GLint ToParam(MinMagFilter minmag_filter,
 }
 
-static GLint ToAddressMode(SamplerAddressMode mode) {
+static GLint ToAddressMode(SamplerAddressMode mode,
```
nit: I might just pass in a bool for whether or not decal is supported.

chinmaygarde replied:
Just a drive by comment; Is it just me or does anyone else find "decal" mode to be a non-standard and/or overly confusing term? How about just `ClampToBorderColor`? Thats what the various APIs call it (`GL_CLAMP_TO_BORDER_EXT`, `MTLSamplerAddressModeClampToBorderColor`, `VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_BORDER`).

Perhaps later, we can rename this enum value for consistency.

jonahwilliams replied:
There is a cost for non-consistency with the underlying graphics API, but there is also a value in matching the framework level enumeration which uses decal as a terminology. I don't feel as strongly about decal as I do about "FilterQuality", which is entirely opaque - at least decal has some semantic meaning

ColdPaleLight replied:
> nit: I might just pass in a bool for whether or not decal is supported.

Done

jonahwilliams reviewed (commented):
Nice!

There are a few places where we ifdef in the shader, such as

https://github.com/flutter/engine/blob/a18c52768bbc8c779604a186dc6f9d14e73cee14/impeller/entity/shaders/blending/porter_duff_blend.frag#L32

I think we should be able to update the ifdef to check for the extension? or something like that.

chinmaygarde commented on impeller/renderer/backend/gles/test/capabilities_unittests.cc:
```diff
@@ -34,5 +34,15 @@ TEST(CapabilitiesGLES, CanInitializeWithDefaults) {
 
+TEST(CapabilitiesGLES, SupportsDecalSamplerAddressMode) {
+  auto const extensions = std::vector<unsigned char*>{
+      (unsigned char*)"GL_KHR_debug",                 //
```
Nit: `reinterpret_cast`

ColdPaleLight replied:
Done

chinmaygarde commented on impeller/renderer/backend/gles/sampler_gles.cc:
```diff
@@ -96,10 +100,12 @@ bool SamplerGLES::ConfigureBoundTexture(const TextureGLES& texture,
-                   ToAddressMode(desc.width_address_mode));
-  gl.TexParameteri(target.value(), GL_TEXTURE_WRAP_T,
-                   ToAddressMode(desc.height_address_mode));
+  gl.TexParameteri(
```
How do we set the border color? `TEXTURE_BORDER_COLOR_EXT`

jonahwilliams replied:
Maybe this defaults to transparent black so we don't need to worry about it?

jonahwilliams replied:
according to https://registry.khronos.org/OpenGL-Refpages/es3/html/glTexParameter.xhtml this value defaults to transparent black so we don't need to set anything.

chinmaygarde replied:
I guess. I took a look at Vulkan and [setting the border color is an extension](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_custom_border_color.html). But setting it to `VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_BORDER` is not. I guess this is fine. I don't think we set the color on Metal either. Just really odd.

jonahwilliams replied:
Flutter does not expose an API for setting the border color.

chinmaygarde replied:
Nevermind. Its an extension that is core in 1.1. And we do set it to [transparent black](https://github.com/flutter/engine/blob/a18c52768bbc8c779604a186dc6f9d14e73cee14/impeller/renderer/backend/vulkan/sampler_library_vk.cc#L46). Let's be explicit and do the same here as well please?

chinmaygarde replied:
Its is also dealers choice on Metal at the moment. Let's be explicit and set borderColor [here too](https://github.com/flutter/engine/blob/a18c52768bbc8c779604a186dc6f9d14e73cee14/impeller/renderer/backend/metal/sampler_library_mtl.mm#L25C17-L25C37). In fact, a default border color specification in `formats.h` would be good for consistency.

chinmaygarde replied:
> Flutter does not expose an API for setting the border color.

Yeah, just didn't want to read the spec to see if the default was transparent black. Just being explicit seemed better. There isn't any bug though.

ColdPaleLight replied:
Issue filed: https://github.com/flutter/flutter/issues/136153

ColdPaleLight commented:
> Nice!
> 
> There are a few places where we ifdef in the shader, such as
> 
> https://github.com/flutter/engine/blob/a18c52768bbc8c779604a186dc6f9d14e73cee14/impeller/entity/shaders/blending/porter_duff_blend.frag#L32
> 
> I think we should be able to update the ifdef to check for the extension? or something like that.

I removed the macro and added a variable in the uniform block to determine if the gpu api supports clip to border. (I originally planned to use different pipelines that support decal and those that do not, but I found that this would double the number of advanced blend pipelines.)

jonahwilliams commented:
Rather than use a uniform, can we use a combination of defines - if opengles and at least one of the decal extensions not defined? Each extension should have a corresponding define already

ColdPaleLight commented:
> Rather than use a uniform, can we use a combination of defines - if opengles and at least one of the decal extensions not defined? Each extension should have a corresponding define already

I guess this won't work because shaderc will do a preprocessing before compiling glsl, so the macros will be handled at that stage. Therefore, glsl generated by spirv cross no longer contains statements like "#ifdef GL_EXT_texture_border_clamp"

jonahwilliams commented:
that is extremely unfortunate

jonahwilliams commented:
Well, we can clean it up later then. There is probably another way to do it by way of tinkering with SPIRV cross.

jonahwilliams commented on impeller/entity/shaders/blending/advanced_blend.glsl:
```diff
@@ -24,12 +25,11 @@ in vec2 v_src_texture_coords;
-#else
-  return texture(texture_sampler, texture_coords);
-#endif
+  if (blend_info.supports_decal_sampler_address_mode > 0.0) {
```
These can still technically be ifdefs since vulkan/metal always support this.

ColdPaleLight replied:
Got it, added them back.

jonahwilliams reviewed (approved):
LGTM
